Implement cell tracking module

dnerini opened the pull request:
This PR implements the Thunderstorm Detection and Tracking (DATing) module , following the procedures used in the TRT Thunderstorms Radar Tracking algorithm (Hering et al 2004) used operationally at MeteoSwiss.

Modifications include advecting the identified thunderstorms with the optical flow obtained from pysteps, as well as additional options in the thresholding.

RTD
----
- feature doc: https://pysteps.readthedocs.io/en/cell_tracking/pysteps_reference/feature.html
- tracking https://pysteps.readthedocs.io/en/cell_tracking/pysteps_reference/tracking.html
- example: https://pysteps.readthedocs.io/en/cell_tracking/auto_examples/thunderstorm_detection_and_tracking.html

PR progress 
-------------
- [x] Create new `feature` and `tracking` modules, including the `__init__.py` and `interface.py` files
- [x] Move existing feature detection methods from `pysteps.utils.images` to `pysteps.feature`
- [x] Move LK tracking method from `pysteps.motion.lucaskanade.track_features` to `pysteps.tracking.lucaskanade.track_features`
- [x] Check requirements
- [x] Add example
- [x] Update documentation
- [x] Add/fix tests

dnerini committed:
Create detection module

dnerini committed:
Create tracking module

dnerini committed:
Migrate detection methods

dnerini committed:
Migrate tracking methods

dnerini committed:
Rename 'detection' to 'feature'

dnerini committed:
Update new __init__.py files

dnerini committed:
Update and fix interface testing

dnerini committed:
Fix references

feldmann-m committed:
File initialization
* Initialized all files for thunderstorm tracking
* Ran black over entire pysteps directory

feldmann-m committed:
Filled initialized files with pre-tested code
* added code to example, detection, tracking and visualization
* so far unformatted with black
* needs to be tested within pysteps environment

feldmann-m committed:
Formatted with black
* all files were processed with black
* in this configuration the code works without errors in the local setup
* path specifications still need to be generalized

feldmann-m committed:
Generalization of paths
* generalizing path of example data
* path of config file is still explicit

dnerini committed:
No need to download the data in examples

dnerini committed:
Return figure axes instead of saving it

dnerini committed:
Add requirements for thunderstorm example

dnerini commented:
Hello @feldmann-m

commits 2ffb21f, 1ddb3b3 and bb3a252  are necessary to render your example in readthedocs. 
(it's building now, let's see if it succeeds).

In particular, in 1ddb3b3 I removed from your plotting functions the part where it saved the figures. Both routine return the figure axes instead. 

In 2ffb21f I remove the downloading of the example data. This is not necessary, since the data are downloaded once for all the examples before readthedocs is build.

dnerini committed:
Update feature doc and interface

dnerini commented on pysteps/feature/tstorm.py:
```diff
@@ -0,0 +1,224 @@
+def get_profile(areas, binary, ref, loc_max, time, minref):
+    """
+    This function returns the identified cells in a dataframe including their x,y
+    locations, location of their maxima, maximum reflectivity and contours.
```
Here `x` and `y` are in unit of number of pixels from the field origin, correct?
This might create some confusion later on when, for example, one would try to plot the contours.

Should we instead work in geographical units? All pysteps data must come on a regular grid and the metadata dictionary will include the definition of the geographical projection.

dnerini replied:
At least partially addressed in f0ebdb4

dnerini committed:
Update tracking doc and interface

dnerini committed:
Plot precip field with pysteps

dnerini committed:
Update requirements for cell tracking example

dnerini committed:
Fix module description

dnerini committed:
Update interface tests with new methods

dnerini committed:
Fix module docstrings

dnerini committed:
Sort imports

dnerini committed:
Rename module 'thunderstorms' for better consistency with existing modules

feldmann-m committed:
Dynamic lower threshold scheme in cell detection
* evaluation of reflectivity along joint cell boundaries
* constrain cell to smaller contour at maxref of boundary
* optional tool, default is fixed lower threshold
* size of cells needs to be controlled, work in progress

dnerini committed:
Remove wrong doi (#185)

feldmann-m committed:
Dynamic lower reflectivity threshold in cell detection
* control of cell size
* control of fragmentization

dnerini committed:
Merge branch 'master' into cell_tracking

dnerini committed:
Use small caps

dnerini committed:
Add cartopy

dnerini committed:
Convert to geographical coordinates

dnerini committed:
Add requirements for cartopy

dnerini committed:
Use conda instead of pip

dnerini committed:
Fix conda package names

dnerini committed:
Remove channel

dnerini committed:
Do not name conda env

dnerini committed:
Revert to pip installation

feldmann-m committed:
Updated Example
* added comments
* filtered tracks to improve plot

feldmann-m committed:
Added MRMS data to example
* example with US data
* size threshold adapted
* noticed issue with adaptive threshold labelling, to be investigated in next update
* examples use fixed threshold for now

feldmann-m committed:
subpress variable threshold

dnerini committed:
Merge branch 'master' into cell_tracking

dnerini committed:
Install pygrib to use mrms data in the examples

dnerini committed:
Merge branch 'master' into cell_tracking

dnerini committed:
Revert "Install pygrib to use mrms data in the examples"

This reverts commit e18a0fa3f7f34e3bf0502c5a3bd5f6fb61d4e191.

dnerini commented on examples/thunderstorm_detection_and_tracking.py:
```diff
@@ -0,0 +1,199 @@
+
+#%%
+################################################################################
+# Example with US MRMS data
```
This is a great idea, @feldmann-m , but unfortunately loading the mrms data requires the pygrib package which is a bit problematic to install without conda (which we don't use on readthedocs). Also, the US data are somewhat heavy and might slow down the build of the documentation or event stop it due to time or memory limits.

I suggest that you only use Swiss data in this example, while the mrms dataset could become a more in-depth tutorial that we can publish as a Google Colab notebook. What do you think?

feldmann-m committed:
Work on example
* split example into 2, US example to be moved to google collab
* added temporal development graphs
* fixed bug in dynamic thresholding

feldmann-m committed:
Documentation and feature detection
* added documentation to all functions
* added reference to TRT
* modified reference.bib file to contain TRT reference
* modified output of cell detection: if output_feat=True in the input variables, the function will return a 2-D array of cell centroids that can be used for flow calculations. The default is false, in that case the dataframes of the cells and their label-matrix are returned

feldmann-m committed:
Ran black over all files

dnerini committed:
Update docstrings

dnerini committed:
Add option to use tstorm as feature detection

dnerini committed:
Local approach if k is None OR coord is None

dnerini committed:
Test LK with different feature detection methods

dnerini committed:
More testing

dnerini committed:
Account for optional dependencies

dnerini committed:
Update test dependencies

feldmann-m committed:
removed extensive example

feldmann-m committed:
Merge branch 'cell_tracking' of https://github.com/pySTEPS/pysteps into cell_tracking

feldmann-m committed:
reformatted example

dnerini committed:
Avoid [] default arguments, plus some additional checks on the arguments

dnerini commented:
@feldmann-m if you were wondering why we need to avoid empty lists as default argument to a function (bcb898e), you can have a look here at the resulting documentation (see screenshot below).
If you interested to better understand the cause, you can have a look here:
https://stackoverflow.com/questions/1132941/least-astonishment-and-the-mutable-default-argument

![image](https://user-images.githubusercontent.com/11967971/100353493-db209680-2fee-11eb-99c0-b4d43c381bb7.png)

feldmann-m commented:
> @feldmann-m if you were wondering why we need to avoid empty lists as default argument to a function ([bcb898e](https://github.com/pySTEPS/pysteps/commit/bcb898ed99ba4a8a2969ef85c975c24c67192621)), you can have a look here at the resulting documentation (see screenshot below).
> If you interested to better understand the cause, you can have a look here:
> https://stackoverflow.com/questions/1132941/least-astonishment-and-the-mutable-default-argument

Thank you for the advice! I will keep that in mind in future. Should I change this in the example as well?

dnerini commented:
> 
> 
> > @feldmann-m if you were wondering why we need to avoid empty lists as default argument to a function ([bcb898e](https://github.com/pySTEPS/pysteps/commit/bcb898ed99ba4a8a2969ef85c975c24c67192621)), you can have a look here at the resulting documentation (see screenshot below).
> > If you interested to better understand the cause, you can have a look here:
> > https://stackoverflow.com/questions/1132941/least-astonishment-and-the-mutable-default-argument
> 
> Thank you for the advice! I will keep that in mind in future. Should I change this in the example as well?

No, no need to change anything in the example script. This applies only when defining the default arguments in a function.

dnerini committed:
Fix refernce and docstring cosmetics

dnerini committed:
Basic testig of cell tracking modules

dnerini committed:
Attempt to fix Panda's SettingWithCopyWarning

dnerini commented:
@feldmann-m can you please have a look at commit 97fb795? It seems to remove the SettingWithCopyWarnings, but would be better to double check that this did not change the behavior of your code.

RubenImhoff commented:
Thanks for the amazing work, @feldmann-m and @dnerini! I'll try to go through the code today. If you think the code is ready for testing, I might give it a try with the KNMI radar data.

RubenImhoff commented on pysteps/feature/tstorm.py:
```diff
@@ -0,0 +1,320 @@
+    :toctree: ../generated/
+
+    detection
+    belonging
```
In the ReadTheDocs, there is no description of 'belonging'. The other three (detection, longdistance and get_profile) do have a description.

dnerini replied:
8ad3af4 should have fixed this

RubenImhoff commented on pysteps/tracking/lucaskanade.py:
```diff
@@ -0,0 +1,201 @@
+    :toctree: ../generated/
+
+    track_features
+"""
```
Perhaps good to mention (for new users) here and/or in motion.lucaskanade.py that the same optical flow algorithm is used, but that the tracking functionality is defined in pysteps.tracking and vice versa for the standard motion derivation.

aperezhortal replied:
Also, it can be briefly mentioned the differences between the two. To prevent any confusion.

RubenImhoff committed:
Minor changes in the comments

RubenImhoff committed:
Update thunderstorm_detection_and_tracking.py

I added some extra comment lines and I made some minor changes to the final two plots. Another idea, but this may just be looking for more work: would it be possible to link the track colours in the first figure to the colour of the tracks in the two evaluation figures? Other than that, nice example! :)

RubenImhoff committed:
Keep references in same structure

RubenImhoff commented on pysteps/feature/blob.py:
```diff
@@ -0,0 +1,99 @@
+        Array of shape (m, n) containing the input image. Nan values are ignored.
+    method : {'log', 'dog', 'doh'}, optional
+        The method to use: 'log' = Laplacian of Gaussian, 'dog' = Difference of
+        Gaussian, ''
```
What about 'doh'?

dnerini replied:
good question, we should ask @pulkkins! But I assume these all come from the scikit-image documentation...

edit: doh = Determinant of Hessians (b68318f)

RubenImhoff committed:
Small comment changes in blob.py

dnerini committed:
Update dev requirements

dnerini committed:
More testing

RubenImhoff commented on pysteps/feature/tstorm.py:
```diff
@@ -0,0 +1,320 @@
+            labels[labels == n] = 0
+        if maxval < minmax:
+            binary[labels == n] = 0
+            labels[labels == n] = 0
```
Perhaps good to add some comments in these lines.

feldmann-m replied:
I added descriptions to lines 137 and 140 to describe the steps being done here.

dnerini committed:
Reformat tests with Black

dnerini committed:
Complete docstrings

- doh = Determinant of Hessian

dnerini committed:
Merge branch 'master' into cell_tracking

loforest commented on pysteps/feature/tstorm.py:
```diff
@@ -0,0 +1,320 @@
+        Minimum distance between two maxima of identified objects. Objects with a
+        smaller distance will be merged. The default is 10 km.
+    dyn_thresh: bool, optional
+        Set to True to activate dynamic lower threshold. Restricts contours to more
```
When the dynamic threshold is activated, does it override the minmax or minref value? 
It is not clear how it operates.

feldmann-m replied:
In general it overrides the minref value, however this module has been removed for now, as I found additional issues.

loforest commented on pysteps/feature/tstorm.py:
```diff
@@ -0,0 +1,320 @@
+            "pandas is required for thunderstorm DATing " "but it is not installed"
+        )
+    filt_image = np.zeros(input_image.shape)
+    filt_image[input_image > minref] = input_image[input_image > minref]
```
According to the definition of minref in the help, the comparison operator should be equal or greater than minref. Correct?

feldmann-m replied:
I corrected it to >= in the code in line 120.

RubenImhoff committed:
Update lucaskanade.py

RubenImhoff commented on pysteps/tracking/tdating.py:
```diff
@@ -0,0 +1,309 @@
+    timelist : list
+        List of length t containing string of time and date of each (m,n) field.
+    mintrack : int, optional
+        minimum length of cell-track to be counted. The default is 3.
```
What is the dimension, pixels?

feldmann-m replied:
The dimension for mintrack is number of timesteps, I expanded the docstrings accordingly.

dnerini committed:
Update autosummary

pulkkins committed:
Docstring polishing

pulkkins committed:
Merge branch 'cell_tracking' of https://github.com/pySTEPS/pysteps into cell_tracking

pulkkins committed:
Use shorter form of boolean comparison

pulkkins commented on pysteps/tracking/tdating.py:
```diff
@@ -0,0 +1,309 @@
+        tracking from there. The default is 0, which initiates a new tracking sequence.
+    minref : float, optional
+        Lower threshold for object detection. Lower values will be set to NaN.
+        The default is 35 dBZ.
```
This might confuse some users. Default values are specified in dBZ. However, the expected unit of the input data is not specified anywhere. I guess it can take any input field regardless of the unit?

pulkkins replied:
To me it looks like minref (and several other keyword arguments) are not used anywhere in the code.

feldmann-m replied:
I expanded the description of the default input.
Thank you very much for pointing out the lack of use of the keywords. This was a mistake in the code, that has now been corrected. Unfortunately this revealed some ongoing issues in the dynthresh function, so I removed it for now.

RubenImhoff commented on pysteps/visualization/thunderstorms.py:
```diff
@@ -0,0 +1,126 @@
+
+
+def plot_track(track_list, geodata=None):
+    """
```
Shouldn't the function also contain xsize, ysize, poix and poiy, or is this part of 'geodata'?

dnerini replied:
Well spotted, I need to update the docstring of this function!

dnerini replied:
eaa2813

RubenImhoff committed:
Update thunderstorms.py

RubenImhoff committed:
Remove module imports that are not used in examples

RubenImhoff reviewed (commented):
And I finally got through all the code. :) Very nice work, @feldmann-m and @dnerini. I left a couple of small comments or questions, and I made some minor changes to the code (mostly docstring related). I have just tested the functionality with the KNMI radar data and it works, see also the figure below!

![Tracks](https://user-images.githubusercontent.com/31476760/100614105-e557d300-3315-11eb-9d2b-940c352f74da.png)

dnerini committed:
Update docstrings

aperezhortal commented on pysteps/feature/tstorm.py:
```diff
@@ -0,0 +1,321 @@
+
+    Returns
+    -------
+    cells_id : pandas dataframe
```
Perhaps we can describe in more detail the contents (columns) of the dataframe. This will make clear what this function returns.

feldmann-m replied:
I added a description of the dataframe columns to the docstrings.

feldmann-m committed:
Comments from review
*description of dataframe columns
*added standard input description
*correction on minref
*expanded explanation of dynthresh
*added comments in tstorm
*remaining issues in dynthresh detected

feldmann-m committed:
formatted with black

feldmann-m committed:
Removed dyn_thresh procedure
*remaining bugs in dyn_thresh need more work
*material for future update

feldmann-m committed:
formatted with black

feldmann-m committed:
minor edit in example

dnerini committed:
Fix tests and typo

dnerini commented:
This now looks to me pretty ready to be merged, what do you think, @feldmann-m ? Something's still missing?

aperezhortal commented:
Great contribution @dnerini and @feldmann-m!  

There are a few minor comments still unresolved. Other than that, it looks ready to go!

feldmann-m commented:
> And I finally got through all the code. :) Very nice work, @feldmann-m and @dnerini. I left a couple of small comments or questions, and I made some minor changes to the code (mostly docstring related). I have just tested the functionality with the KNMI radar data and it works, see also the figure below!

Thank you very much for testing it on the KNMI data! It is great to see, that it works on other data as well. I also ran a test on MRMS data, which seems for work fine as well.

feldmann-m commented:
> Great contribution @dnerini and @feldmann-m!
> 
> There are a few minor comments still unresolved. Other than that, it looks ready to go!

I hope I addressed the comments in the updates from Thursday, if there are remaining unresolved issues, please let me know!

dnerini commented:
All right, I'm going to merge this PR, since all pending comments are resolved. 

@feldmann-m thank you so much for the hard work. I'm sure this new module will be highly appreciated and used!
Hopefully this will also inspire some new and exciting developments on thunderstorm detection, tracking, and nowcasting... we'll see! Again, well done!
